src/Symfony/Component/DependencyInjection/Compiler/CheckTypeHintsPass.php

```diff
@@ -0,0 +1,165 @@
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Julien Maulny <jmaulny@darkmira.fr>
+ */
+class CheckTypeHintsPass extends AbstractRecursivePass
+{
+    private $autoload;
+
+    public function __construct(bool $autoload = false)
+    {
+        $this->autoload = $autoload;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        // When service is declared with $container->register(Foo::class); and not $container->register(Foo::class, Foo::class);
+        if (null === $value->getClass()) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        if (!$this->autoload && !\class_exists($this->getClassName($value), false)) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        if (null !== $constructor = $this->getConstructor($value, false)) {
+            $this->checkArgumentsTypeHints($constructor, $value->getArguments());
+        }
+
+        foreach ($value->getMethodCalls() as $methodCall) {
+            $reflectionMethod = $this->getReflectionMethod($value, $methodCall[0]);
+
+            $this->checkArgumentsTypeHints($reflectionMethod /* warn */, $methodCall[1]);
+        }
+
+        return parent::processValue($value, $isRoot);
+    }
+
+    /**
+     * Check type hints for every parameter of a method/constructor.
+     *
+     * @param \ReflectionMethod $reflectionMethod
+     * @param Reference[]       $configurationArguments
+     */
+    private function checkArgumentsTypeHints(\ReflectionMethod $reflectionMethod, array $configurationArguments)
+    {
+        $numberOfRequiredParameters = $reflectionMethod->getNumberOfRequiredParameters();
+
+        if (count($configurationArguments) < $numberOfRequiredParameters) {
+            throw new InvalidArgumentException(sprintf(
+                'In service declaration "%s". Missing parameter in "%s::%s". Expecting %d required parameters, got only %d.',
+                $this->currentId,
+                $reflectionMethod->getDeclaringClass()->getName(),
+                $reflectionMethod->getName(),
+                $numberOfRequiredParameters,
+                count($configurationArguments)
+            ));
+        }
+
+        $reflectionParameters = $reflectionMethod->getParameters();
+        $checksCount = min($reflectionMethod->getNumberOfParameters(), count($configurationArguments));
+
+        if ($reflectionMethod->isVariadic()) {
+            --$checksCount;
+        }
+
+        for ($i = 0; $i < $checksCount; ++$i) {
+            if (!$reflectionParameters[$i]->hasType()) {
+                continue;
+            }
+
+            $this->checkTypeHint($configurationArguments[$i], $reflectionParameters[$i]);
+        }
+
+        if ($reflectionMethod->isVariadic() && ($lastParameter = end($reflectionParameters))->hasType()) {
```

Comment:
brackets around $lastParameter... should be removed

Reply:
I can't just remove brackets, I'm using `()->hasType()` after the output of brackets